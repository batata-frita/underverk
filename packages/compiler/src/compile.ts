import * as t from '@babel/types'
import template from '@babel/template'
import {
  Component,
  State,
  Expression,
  Declaration,
  Effect,
  Child,
  Literal,
  Computed,
  Reference,
  Function,
  UpdateFunction,
  Argument,
  Context,
  LiteralValue,
  ContextGetter,
  ContextSetter,
  Project,
} from './types'

export const compile = (projectAst: Project): t.VariableDeclaration[] => [
  ...projectAst.contexts.map(compileContext),
  ...projectAst.components.map(compileComponent),
]

export const compileContext = (contextAst: Context): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(contextAst.name),
      t.callExpression(t.identifier('createContext'), [compileLiteralValue(contextAst.defaultValue)]),
    ),
  ])

export const compileComponent = (componentAst: Component): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(componentAst.name),
      t.arrowFunctionExpression(
        [compileProps(componentAst.props)],
        t.blockStatement([
          ...componentAst.literals.map(compileLiteral),
          ...componentAst.getContexts.map(compileGetContext),
          ...componentAst.functions.map(compileFunction(componentAst.props)),
          ...(componentAst.state ? compileState(componentAst.state) : []),
          ...componentAst.computed.map(compileComputed),
          ...componentAst.effects.map(compileEffect),
          t.returnStatement(compileSetContexts(componentAst.setContexts, compileChildren(componentAst.children))),
        ]),
      ),
    ),
  ])

const compileGetContext = (getContextAst: ContextGetter): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(getContextAst.name),
      t.callExpression(t.identifier('useContext'), [t.identifier(getContextAst.context)]),
    ),
  ])

const compileSetContexts = (
  setContextsAst: ContextSetter[],
  children: t.JSXElement | t.JSXFragment,
): t.JSXElement | t.JSXFragment => {
  if (setContextsAst.length === 0) {
    return children
  }

  const [head, ...tail] = setContextsAst
  return compileSetContexts(
    tail,
    t.jsxElement(
      t.jsxOpeningElement(t.jsxIdentifier(`${head.context}.Provider`), [
        t.jsxAttribute(t.jsxIdentifier('value'), t.jsxExpressionContainer(t.identifier(head.value))),
      ]),
      t.jsxClosingElement(t.jsxIdentifier(`${head.context}.Provider`)),
      [children],
      false,
    ),
  )
}

const compileProps = (propsAst: Argument[]): t.ObjectPattern =>
  t.objectPattern(propsAst.map(prop => t.objectProperty(t.identifier(prop.name), t.identifier(prop.name))))

const compileLiteral = (literalAst: Literal): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(t.identifier(literalAst.name), compileLiteralValue(literalAst.value)),
  ])

const compileLiteralValue = (literalValue: LiteralValue): t.Expression => {
  switch (typeof literalValue) {
    case 'string':
      return t.stringLiteral(literalValue)

    case 'boolean':
      return t.booleanLiteral(literalValue)

    case 'number':
      return t.numericLiteral(literalValue)

    case 'object':
      return template.expression(JSON.stringify(literalValue))()

    case 'undefined':
      return t.identifier('undefined')

    default:
      console.warn('ðŸ™ˆ literal captured by default case', literalValue)
      return t.stringLiteral(`${literalValue}`)
  }
}

const compileFunction = (props: Argument[]) => (functionAst: Function): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(functionAst.name),
      t.callExpression(t.identifier('useCallback'), [
        t.arrowFunctionExpression(
          functionAst.arguments.map(argument => t.identifier(argument.name)),
          compileComposition(functionAst.composition),
        ),
        t.arrayExpression(props.map(({ name }) => t.identifier(name))),
      ]),
    ),
  ])

const compileComposition = (expressionAst: Expression[]): t.CallExpression =>
  t.callExpression(t.callExpression(t.identifier('compose'), expressionAst.map(compileExpression)), [
    t.identifier('event'),
  ])

const compileChildren = (childrenAst: Child[]): t.JSXFragment =>
  t.jsxFragment(t.jsxOpeningFragment(), t.jsxClosingFragment(), childrenAst.map(compileChild))

const compileChild = (childAst: Child): t.JSXElement | t.JSXText | t.JSXExpressionContainer => {
  switch (childAst.type) {
    case 'staticNode':
      const selfClosingTag = childAst.children.length === 0

      return t.jsxElement(
        t.jsxOpeningElement(t.jsxIdentifier(childAst.element), childAst.props.map(compileJsxProp), selfClosingTag),
        selfClosingTag ? undefined : t.jsxClosingElement(t.jsxIdentifier(childAst.element)),
        childAst.children.map(compileChild),
        selfClosingTag,
      )

    case 'dynamicNode':
      return t.jsxExpressionContainer(
        t.callExpression(t.identifier('dynamicNode'), [
          t.identifier(childAst.element),
          t.identifier(childAst.dependency),
        ]),
      )

    default:
      return t.jsxExpressionContainer(compileExpression(childAst))
  }
}

const compileJsxProp = (propAst: Declaration): t.JSXAttribute =>
  t.jsxAttribute(t.jsxIdentifier(propAst.name), t.jsxExpressionContainer(compileExpression(propAst.value)))

const compileEffect = (effectAst: Effect): t.ExpressionStatement =>
  t.expressionStatement(
    t.callExpression(t.identifier('useEffect'), [
      t.arrowFunctionExpression(
        [],
        t.blockStatement([
          t.expressionStatement(
            t.callExpression(t.identifier(effectAst.type), [
              t.identifier(effectAst.dependency),
              t.identifier(effectAst.handler),
            ]),
          ),
        ]),
      ),
      t.arrayExpression([t.identifier(effectAst.dependency)]),
    ]),
  )

const compileComputed = (computedAst: Computed): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(computedAst.name),
      t.callExpression(t.identifier(computedAst.operation), computedAst.arguments.map(compileReference)),
    ),
  ])

const compileExpression = (expressionAst: Expression): t.Expression => {
  switch (expressionAst.type) {
    case 'operation':
      return t.callExpression(t.identifier(expressionAst.name), expressionAst.arguments.map(compileExpression))

    case 'reference':
      return compileReference(expressionAst)

    default:
      console.warn('ðŸ™ˆ expression captured by default case', expressionAst)
      return t.stringLiteral('')
  }
}

const compileReference = (referenceAst: Reference): t.Identifier => t.identifier(referenceAst.value)

const compileState = (stateAst: State): t.VariableDeclaration[] => [
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.arrayPattern([t.identifier('state'), t.identifier('setState')]),
      t.callExpression(t.identifier('useState'), [compileExpression(stateAst.defaultValue)]),
    ),
  ]),
  ...stateAst.updateFunctions.map(compileUpdateFunction),
]

const compileUpdateFunction = (updateFunctionAst: UpdateFunction): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(updateFunctionAst.name),
      t.callExpression(t.identifier('useCallback'), [
        t.arrowFunctionExpression(
          [t.identifier('event')],
          t.callExpression(t.identifier('setState'), [
            t.objectExpression([
              t.spreadElement(t.identifier('state')),
              t.spreadElement(
                t.callExpression(t.identifier(updateFunctionAst.transformation), [
                  t.identifier('event'),
                  t.identifier('state'),
                ]),
              ),
            ]),
          ]),
        ),
        t.arrayExpression([t.identifier('state'), t.identifier(updateFunctionAst.transformation)]),
      ]),
    ),
  ])
