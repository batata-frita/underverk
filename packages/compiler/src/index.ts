import * as t from '@babel/types'
import template from '@babel/template'
import babelGenerate from '@babel/generator'
import {
  Component,
  State,
  Expression,
  Declaration,
  Effect,
  Child,
  Literal,
  Computed,
  Reference,
  Function,
  UpdateFunction,
  Argument,
} from './types'
export * from './types'
export * from './manipulations'

export const generate = (ast: t.VariableDeclaration): string => babelGenerate(ast).code

export const compile = (componentAst: Component): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(componentAst.name),
      t.arrowFunctionExpression(
        [compileProps(componentAst.props)],
        t.blockStatement([
          ...componentAst.literals.map(compileLiteral),
          ...componentAst.functions.map(compileFunction(componentAst.props)),
          ...(componentAst.state ? compileState(componentAst.state) : []),
          ...componentAst.computed.map(compileComputed),
          ...componentAst.effects.map(compileEffect),
          compileChildren(componentAst.children),
        ]),
      ),
    ),
  ])

const compileProps = (propsAst: Argument[]): t.ObjectPattern =>
  t.objectPattern(propsAst.map(prop => t.objectProperty(t.identifier(prop.name), t.identifier(prop.name))))

const compileLiteral = (literalAst: Literal): t.VariableDeclaration => {
  switch (typeof literalAst.value) {
    case 'string':
      return t.variableDeclaration('const', [
        t.variableDeclarator(t.identifier(literalAst.name), t.stringLiteral(literalAst.value)),
      ])

    case 'boolean':
      return t.variableDeclaration('const', [
        t.variableDeclarator(t.identifier(literalAst.name), t.booleanLiteral(literalAst.value)),
      ])

    case 'number':
      return t.variableDeclaration('const', [
        t.variableDeclarator(t.identifier(literalAst.name), t.numericLiteral(literalAst.value)),
      ])

    case 'object':
      return t.variableDeclaration('const', [
        t.variableDeclarator(t.identifier(literalAst.name), template.expression(JSON.stringify(literalAst.value))()),
      ])

    case 'undefined':
      return t.variableDeclaration('let', [t.variableDeclarator(t.identifier(literalAst.name))])

    default:
      console.warn('ðŸ™ˆ literal captured by default case', literalAst)
      return t.variableDeclaration('const', [
        t.variableDeclarator(t.identifier(literalAst.name), t.stringLiteral(`${literalAst.value}`)),
      ])
  }
}

const compileFunction = (props: Argument[]) => (functionAst: Function): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(functionAst.name),
      t.callExpression(t.identifier('useCallback'), [
        t.arrowFunctionExpression(
          functionAst.arguments.map(argument => t.identifier(argument.name)),
          compileComposition(functionAst.composition),
        ),
        t.arrayExpression(props.map(({ name }) => t.identifier(name))),
      ]),
    ),
  ])

const compileComposition = (expressionAst: Expression[]): t.CallExpression =>
  t.callExpression(t.callExpression(t.identifier('compose'), expressionAst.map(compileExpression)), [
    t.identifier('event'),
  ])

const compileChildren = (childrenAst: Child[]): t.ReturnStatement =>
  t.returnStatement(t.jsxFragment(t.jsxOpeningFragment(), t.jsxClosingFragment(), childrenAst.map(compileChild)))

const compileChild = (childAst: Child): t.JSXElement | t.JSXText | t.JSXExpressionContainer => {
  switch (childAst.type) {
    case 'staticNode':
      const selfClosingTag = childAst.children.length === 0

      return t.jsxElement(
        t.jsxOpeningElement(t.jsxIdentifier(childAst.element), childAst.props.map(compileJsxProp), selfClosingTag),
        selfClosingTag ? undefined : t.jsxClosingElement(t.jsxIdentifier(childAst.element)),
        childAst.children.map(compileChild),
        selfClosingTag,
      )

    case 'dynamicNode':
      return t.jsxExpressionContainer(
        t.callExpression(t.identifier('dynamicNode'), [
          t.identifier(childAst.element),
          t.identifier(childAst.dependency),
        ]),
      )

    default:
      return t.jsxExpressionContainer(compileExpression(childAst))
  }
}

const compileJsxProp = (propAst: Declaration): t.JSXAttribute =>
  t.jsxAttribute(t.jsxIdentifier(propAst.name), t.jsxExpressionContainer(compileExpression(propAst.value)))

const compileEffect = (effectAst: Effect): t.ExpressionStatement =>
  t.expressionStatement(
    t.callExpression(t.identifier('useEffect'), [
      t.arrowFunctionExpression(
        [],
        t.blockStatement([
          t.expressionStatement(
            t.callExpression(t.identifier(effectAst.type), [
              t.identifier(effectAst.dependency),
              t.identifier(effectAst.handler),
            ]),
          ),
        ]),
      ),
      t.arrayExpression([t.identifier(effectAst.dependency)]),
    ]),
  )

const compileComputed = (computedAst: Computed): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(computedAst.name),
      t.callExpression(t.identifier(computedAst.operation), computedAst.arguments.map(compileReference)),
    ),
  ])

const compileExpression = (expressionAst: Expression): t.Expression => {
  switch (expressionAst.type) {
    case 'operation':
      return t.callExpression(t.identifier(expressionAst.name), expressionAst.arguments.map(compileExpression))

    case 'reference':
      return compileReference(expressionAst)

    default:
      console.warn('ðŸ™ˆ expression captured by default case', expressionAst)
      return t.stringLiteral('')
  }
}

const compileReference = (referenceAst: Reference): t.Identifier => t.identifier(referenceAst.value)

const compileState = (stateAst: State): t.VariableDeclaration[] => [
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.arrayPattern([t.identifier('state'), t.identifier('setState')]),
      t.callExpression(t.identifier('useState'), [compileExpression(stateAst.defaultValue)]),
    ),
  ]),
  ...stateAst.updateFunctions.map(compileUpdateFunction),
]

const compileUpdateFunction = (updateFunctionAst: UpdateFunction): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(updateFunctionAst.name),
      t.callExpression(t.identifier('useCallback'), [
        t.arrowFunctionExpression(
          [t.identifier('event')],
          t.callExpression(t.identifier('setState'), [
            t.objectExpression([
              t.spreadElement(t.identifier('state')),
              t.spreadElement(
                t.callExpression(t.identifier(updateFunctionAst.transformation), [
                  t.identifier('event'),
                  t.identifier('state'),
                ]),
              ),
            ]),
          ]),
        ),
        t.arrayExpression([t.identifier('state'), t.identifier(updateFunctionAst.transformation)]),
      ]),
    ),
  ])
