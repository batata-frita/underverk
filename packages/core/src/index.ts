import * as t from '@babel/types'
import { Component, State, Expression, Declaration } from './types'

export default (componentAst: Component): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(componentAst.name),
      t.arrowFunctionExpression(
        [],
        t.blockStatement([
          ...componentAst.states.map(compileState),
          ...componentAst.declarations.map(compileDeclaration),
        ]),
      ),
    ),
  ])

const compileDeclaration = (declarationAst: Declaration): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(t.identifier(declarationAst.name), compileExpression(declarationAst.value)),
  ])

const compileExpression = (expressionAst: Expression): t.Expression => {
  switch (expressionAst.type) {
    case 'literal':
      return t.stringLiteral(`${expressionAst.value}`)

    case 'function':
      return t.callExpression(t.identifier('callback'), [
        t.arrowFunctionExpression(
          expressionAst.arguments.map(argument => t.identifier(argument.name)),
          compileExpression(expressionAst.body),
        ),
      ])

    case 'operation':
      return t.callExpression(t.identifier(expressionAst.name), expressionAst.arguments.map(compileExpression))

    case 'reference':
      return t.identifier(expressionAst.value)

    default:
      console.warn('ðŸ™ˆ expression captured by default case', expressionAst)
      return t.stringLiteral('')
  }
}

const compileState = (stateAst: State): t.VariableDeclaration =>
  t.variableDeclaration('const', [
    t.variableDeclarator(
      t.arrayPattern([t.identifier(stateAst.name), t.identifier(stateAst.updateFunction)]),
      t.callExpression(t.identifier('state'), [compileExpression(stateAst.defaultValue)]),
    ),
  ])
