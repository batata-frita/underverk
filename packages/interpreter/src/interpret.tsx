import React, { FC, ContextType } from 'react'
import { Component, Literal, Argument, ContextGetter } from './types'

let availableComponents: { [key: string]: FC } = {
  div: (props: {}) => <div {...props} />,
  svg: (props: {}) => <svg {...props} />,
}

export const interpretComponent = (componentAst: Component, contexts: { [key: string]: ContextType<any> }): FC => {
  const Result = (props: {}) => {
    const references: { [key: string]: any } = {
      ...interpretProps(componentAst.props, props),
      ...interpretLiterals(componentAst.literals),
      ...interpretContextGetters(componentAst.getContexts, contexts),
    }

    return (
      <>
        {componentAst.children.map((child, index) => {
          switch (child.type) {
            case 'staticNode': {
              const Target = availableComponents[child.element]
              return <Target key={index} />
            }

            case 'reference':
              return <span key={index}>{references[child.value]}</span>

            case 'dynamicNode':
              return <span key={index}>later</span>
          }
        })}
      </>
    )
  }

  Result.displayName = componentAst.name

  return Result
}

export const interpretProps = (propsAst: Argument[], props: { [key: string]: any }): {} =>
  propsAst.reduce((result, prop) => ({ ...result, [prop.name]: props[prop.name] }), {})

export const interpretLiterals = (literalsAst: Literal[]): {} =>
  literalsAst.reduce((result, literal) => ({ ...result, [literal.name]: literal.value }), {})

export const interpretContextGetters = (
  contextGettersAst: ContextGetter[],
  contexts: { [key: string]: ContextType<any> },
): {} =>
  contextGettersAst.reduce(
    (result, contextGetter) => ({
      ...result,
      [contextGetter.name]: contexts[contextGetter.context],
    }),
    {},
  )
