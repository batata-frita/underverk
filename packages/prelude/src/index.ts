import React, { PropsWithChildren } from 'react'
import * as R from 'ramda'

export const compose = R.compose

export const concat = R.concat

export const when = (condition: boolean, then: any, otherwise: any) => (condition ? then : otherwise)

export const get = R.curry((key, object) => object[key])

export const set = R.curry((key, value, object) => ({
  ...object,
  [key]: value,
}))

export const objectOf = R.curry((key, value) => ({
  [key]: value,
}))

export const append = R.curry((item, list) => [...list, item])

export const constant = (value: any) => () => value

export const dynamicNode = (
  Component: React.FunctionComponent,
  props: PropsWithChildren<{}> | PropsWithChildren<{}>[],
) => {
  if (Array.isArray(props)) {
    return props.map(props =>
      React.createElement(
        Component,
        R.omit(['children'], props),
        ...(Array.isArray(props.children) ? props.children : [props.children]),
      ),
    )
  }

  if (props !== undefined) {
    return React.createElement(
      Component,
      R.omit(['children'], props),
      ...(Array.isArray(props.children) ? props.children : [props.children]),
    )
  }

  return null
}

export const useEffect = React.useEffect

export const useState = React.useState

export const useCallback = React.useCallback

export const useContext = React.useContext

export const createContext = React.createContext

export const log = (item: any, callback: () => void) => {
  console.log(item)
  callback()
}

export const noop = () => { }

export interface API {
  compose: typeof compose
  concat: typeof concat
  when: typeof when
  get: typeof get
  set: typeof set
  objectOf: typeof objectOf
  append: typeof append
  constant: typeof constant
  dynamicNode: typeof dynamicNode
  useEffect: typeof useEffect
  useState: typeof useState
  useCallback: typeof useCallback
  useContext: typeof useContext
  createContext: typeof createContext
  log: typeof log
  noop: typeof noop
}