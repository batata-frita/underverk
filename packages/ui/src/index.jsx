import React from 'react'
import ReactDOM from 'react-dom'
import './index.css'
import App from './App'
/* eslint-disable import/no-webpack-loader-syntax */
import testProgram from '-!@underverk/loader!./testProgram.underverk.js'

console.log('ðŸ§±', testProgram)

ReactDOM.render(<App />, document.getElementById('root'))
