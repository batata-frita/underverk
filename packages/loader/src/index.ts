import { generateProgram } from '@underverk/compiler'

export default function(jsonSource: string) {
  return generateProgram(JSON.parse(jsonSource))
}
