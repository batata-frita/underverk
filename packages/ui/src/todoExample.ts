import { Component, Project } from '@underverk/compiler'

export const App: Component = {
  name: 'App',
  props: [],
  literals: [
    { name: 'valueKey', value: 'value' },
    { name: 'targetKey', value: 'target' },
    { name: 'listKey', value: 'list' },
    { name: 'currentKey', value: 'current' },
    { name: 'doneKey', value: 'done' },
    { name: 'falseKey', value: 'false' },
    { name: 'defaultCurrent', value: '' },
    { name: 'textKey', value: 'text' },
    { name: 'defaultState', value: { list: [], current: '' } },
    { name: 'submitLabel', value: 'Submit' },
  ],
  functions: [
    {
      name: 'getValueFromEvent',
      arguments: [{ name: 'event' }, { name: 'state' }],
      composition: [
        { type: 'operation', name: 'objectOf', arguments: [{ type: 'reference', value: 'currentKey' }] },
        { type: 'operation', name: 'get', arguments: [{ type: 'reference', value: 'valueKey' }] },
        { type: 'operation', name: 'get', arguments: [{ type: 'reference', value: 'targetKey' }] },
      ],
    },
    {
      name: 'addTodo',
      arguments: [{ name: 'event' }, { name: 'state' }],
      composition: [
        {
          type: 'operation',
          name: 'set',
          arguments: [
            { type: 'reference', value: 'listKey' },
            {
              type: 'operation',
              name: 'append',
              arguments: [
                {
                  type: 'operation',
                  name: 'set',
                  arguments: [
                    { type: 'reference', value: 'doneKey' },
                    { type: 'reference', value: 'falseKey' },
                    {
                      type: 'operation',
                      name: 'objectOf',
                      arguments: [
                        {
                          type: 'reference',
                          value: 'textKey',
                        },
                        {
                          type: 'operation',
                          name: 'get',
                          arguments: [
                            { type: 'reference', value: 'currentKey' },
                            { type: 'reference', value: 'state' },
                          ],
                        },
                      ],
                    },
                  ],
                },
                {
                  type: 'operation',
                  name: 'get',
                  arguments: [
                    { type: 'reference', value: 'listKey' },
                    { type: 'reference', value: 'state' },
                  ],
                },
              ],
            },
          ],
        },
        {
          type: 'operation',
          name: 'set',
          arguments: [
            { type: 'reference', value: 'currentKey' },
            { type: 'reference', value: 'defaultCurrent' },
          ],
        },
        { type: 'operation', name: 'constant', arguments: [{ type: 'reference', value: 'state' }] },
      ],
    },
  ],

  state: {
    defaultValue: { type: 'reference', value: 'defaultState' },
    updateFunctions: [
      { name: 'handleCurrentChange', transformation: 'getValueFromEvent' },
      { name: 'updateClick', transformation: 'addTodo' },
    ],
  },

  computed: [
    {
      name: 'list',
      operation: 'get',
      arguments: [
        { type: 'reference', value: 'listKey' },
        { type: 'reference', value: 'state' },
      ],
    },
    {
      name: 'value',
      operation: 'get',
      arguments: [
        { type: 'reference', value: 'currentKey' },
        { type: 'reference', value: 'state' },
      ],
    },
  ],

  effects: [
    {
      type: 'log',
      dependency: 'list',
      handler: 'noop',
    },
  ],

  children: [
    {
      type: 'staticNode',
      element: 'input',
      props: [
        { name: 'onChange', value: { type: 'reference', value: 'handleCurrentChange' } },
        { name: 'value', value: { type: 'reference', value: 'value' } },
      ],
      children: [],
    },
    {
      type: 'staticNode',
      element: 'button',
      props: [{ name: 'onClick', value: { type: 'reference', value: 'updateClick' } }],
      children: [{ type: 'reference', value: 'submitLabel' }],
    },
    {
      type: 'staticNode',
      element: 'ul',
      props: [],
      children: [
        {
          type: 'dynamicNode',
          element: 'TodoItem',
          dependency: 'list',
        },
      ],
    },
  ],
}

export const TodoItem: Component = {
  name: 'TodoItem',
  props: [
    {
      name: 'done',
    },
    {
      name: 'text',
    },
  ],
  literals: [
    {
      name: 'emptyObject',
      value: {},
    },
    {
      name: 'theUndefined',
      value: undefined,
    },
  ],
  functions: [],
  computed: [
    {
      name: 'checkedProps',
      operation: 'when',
      arguments: [
        {
          type: 'reference',
          value: 'done',
        },
        {
          type: 'reference',
          value: 'emptyObject',
        },
        {
          type: 'reference',
          value: 'theUndefined',
        },
      ],
    },
  ],
  effects: [],
  children: [
    {
      type: 'staticNode',
      element: 'li',
      props: [],
      children: [
        {
          type: 'reference',
          value: 'text',
        },
        {
          type: 'dynamicNode',
          element: 'Checkmark',
          dependency: 'checkedProps',
        },
      ],
    },
  ],
}

export const Checkmark: Component = {
  name: 'Checkmark',
  props: [],
  literals: [
    { name: 'width', value: 20 },
    { name: 'height', value: 20 },
    { name: 'path', value: 'M0,5 L0,5 L5,10 L10,0 Z' },
    { name: 'noneFill', value: 'none' },
    { name: 'strokeWidth', value: '1px' },
    { name: 'stroke', value: 'black' },
  ],
  functions: [],
  computed: [],
  effects: [],
  children: [
    {
      type: 'staticNode',
      element: 'svg',
      props: [
        {
          name: 'width',
          value: { type: 'reference', value: 'width' },
        },
        {
          name: 'height',
          value: { type: 'reference', value: 'height' },
        },
      ],
      children: [
        {
          type: 'staticNode',
          element: 'path',
          props: [
            {
              name: 'fill',
              value: {
                type: 'reference',
                value: 'noneFill',
              },
            },
            {
              name: 'strokeWidth',
              value: {
                type: 'reference',
                value: 'strokeWidth',
              },
            },
            {
              name: 'stroke',
              value: {
                type: 'reference',
                value: 'stroke',
              },
            },
            {
              name: 'd',
              value: {
                type: 'reference',
                value: 'path',
              },
            },
          ],
          children: [],
        },
      ],
    },
  ],
}

export const project: Project = {
  name: 'Todo',
  components: [App, TodoItem, Checkmark],
}
