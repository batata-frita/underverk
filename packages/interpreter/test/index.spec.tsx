import React, { createContext } from 'react'
// @ts-ignore
import renderer from 'react-test-renderer'
import { interpretComponent } from '../src'
import { Checkmark } from './examples/todoExample'
import { Button } from './examples/staticExample'

// describe('statefulExample', () => {
//   it('compiles', () => {
//     expect(compileComponent(App)).toMatchSnapshot()
//   })

//   it('generates', () => {
//     expect(generate(compileComponent(App))).toMatchSnapshot()
//     expect(generate(compileComponent(TodoItem))).toMatchSnapshot()
//     expect(generate(compileComponent(Checkmark))).toMatchSnapshot()
//     expect(generate(compileContext(Theme))).toMatchSnapshot()
//     expect(generateProgram(Project)).toMatchSnapshot()
//   })
// })

describe('staticExample', () => {
  it('interprets', () => {
    const ButtonComponent = interpretComponent(Button, { theme: createContext({}) })
    const CheckmarkComponent = interpretComponent(Checkmark, { theme: createContext({}) })

    expect(renderer.create(<ButtonComponent />).toJSON()).toMatchSnapshot()
    expect(renderer.create(<CheckmarkComponent />).toJSON()).toMatchSnapshot()
  })
})
