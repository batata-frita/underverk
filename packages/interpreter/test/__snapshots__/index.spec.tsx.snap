// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`staticExample interprets 1`] = `<div />`;

exports[`staticExample interprets 2`] = `<svg />`;
